File: core/src/main/java/com/vzome/core/zomic/ZomodDirectionNaming.java

package com.vzome.core.zomic;

import java.util.HashMap;
import java.util.Map;

import com.vzome.core.math.symmetry.Axis;
import com.vzome.core.math.symmetry.Direction;
import com.vzome.core.math.symmetry.DirectionNaming;
import com.vzome.core.math.symmetry.Symmetry;


class ZomodDirectionNaming extends DirectionNaming
	{
		private final int[] mMapping;
		
		private final Map<Axis, String> mBackMap = new HashMap<>();
		
		ZomodDirectionNaming( Direction dir, int[] mapping )
		{
			super( dir, dir .getName() );
			mMapping = mapping;
			for ( int i = 0; i < mMapping .length; i++ ) {
			    Axis axis = dir .getAxis( Symmetry.PLUS, mMapping[ i ] );
			    mBackMap .put( axis, "+" + i );
			    axis = dir .getAxis( Symmetry.MINUS, mMapping[ i ] );
			    mBackMap .put( axis, "-" + i );

//			    System .out .println();
//			    for ( int sense = Symmetry.PLUS; sense <= Symmetry.MINUS; sense++ ) {
//			        String name = SIGN[sense] + i;
//			        axis = getAxis( name );
//			        System .out .println( dir .getName() + " " + name + " is orientation: " + SIGN[axis .getSense()] + axis .getOrientation() );
//			    }
			}

//			System .out .println();
//			for ( int sense = Symmetry.PLUS; sense <= Symmetry.MINUS; sense++ )
//			    for ( int i = 0; i < dir .getSymmetry() .getChiralOrder(); i++ )
//			        System .out .println( dir .getName() + " " + SIGN[sense] + i + " is called: " + getName( dir .getAxis( sense, i ) ) );
		}
		
        @Override
		public Axis getAxis( String axisName )
		{
		    int sense = getSign( axisName );
		    int index = getInteger( axisName );
			return getDirection() .getAxis( sense, mMapping[ index ] );
		}
		
        @Override
		public String getName( Axis axis )
		{
		    return mBackMap .get( axis );
		}
	}

File: online/src/worker/legacy/from-java/com/vzome/core/zomic/ZomodDirectionNaming.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { Axis } from "../math/symmetry/Axis.js";
import { Direction } from "../math/symmetry/Direction.js";
import { DirectionNaming } from "../math/symmetry/DirectionNaming.js";
import { Symmetry } from "../math/symmetry/Symmetry.js";

export class ZomodDirectionNaming extends DirectionNaming {
    /*private*/ mMapping: number[];

    /*private*/ mBackMap: java.util.Map<Axis, string>;

    constructor(dir: Direction, mapping: number[]) {
        super(dir, dir.getName());
        if (this.mMapping === undefined) { this.mMapping = null; }
        this.mBackMap = <any>(new java.util.HashMap<any, any>());
        this.mMapping = mapping;
        for(let i: number = 0; i < this.mMapping.length; i++) {{
            let axis: Axis = dir.getAxis$int$int(Symmetry.PLUS, this.mMapping[i]);
            this.mBackMap.put(axis, "+" + i);
            axis = dir.getAxis$int$int(Symmetry.MINUS, this.mMapping[i]);
            this.mBackMap.put(axis, "-" + i);
        };}
    }

    /**
     * 
     * @param {string} axisName
     * @return {Axis}
     */
    public getAxis(axisName: string): Axis {
        const sense: number = this.getSign(axisName);
        const index: number = this.getInteger(axisName);
        return this.getDirection().getAxis$int$int(sense, this.mMapping[index]);
    }

    public getName$com_vzome_core_math_symmetry_Axis(axis: Axis): string {
        return this.mBackMap.get(axis);
    }

    /**
     * 
     * @param {Axis} axis
     * @return {string}
     */
    public getName(axis?: any): string {
        if (((axis != null && axis instanceof <any>Axis) || axis === null)) {
            return <any>this.getName$com_vzome_core_math_symmetry_Axis(axis);
        } else if (axis === undefined) {
            return <any>this.getName$();
        } else throw new Error('invalid overload');
    }
}
ZomodDirectionNaming["__class"] = "com.vzome.core.zomic.ZomodDirectionNaming";
